File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/util/ShareUtilsTest.java
package org.totschnig.myexpenses.test.util;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import com.annimon.stream.Stream;

import org.junit.Assume;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.totschnig.myexpenses.util.AppDirHelper;
import org.totschnig.myexpenses.util.ShareUtils;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import androidx.test.InstrumentationRegistry;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.junit.Assert.assertEquals;

@RunWith(AndroidJUnit4.class)
public class ShareUtilsTest  {

  @Before
  public void setup() {
    Assume.assumeTrue(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP);
  }

  @Test
  public void shouldConvertSingleFileUri() {
    String mimeType = "text/plain";
    Uri testFileUri = AppDirHelper.getAppDir(InstrumentationRegistry.getTargetContext()).createFile(mimeType,"testFile").getUri();
    assertFileScheme(testFileUri);
    List<Uri> fileUris = Collections.singletonList(testFileUri);
    Intent intent = ShareUtils.buildIntent(fileUris, mimeType, null);
    Uri sharedUri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
    assertContentScheme(sharedUri);
  }

  @Test
  public void shouldConvertMultipleFileUris() {
    String mimeType = "text/plain";
    Context context = InstrumentationRegistry.getTargetContext();
    Uri testFile1Uri = AppDirHelper.getAppDir(context).createFile(mimeType,"testFile1").getUri();
    Uri testFile2Uri = AppDirHelper.getAppDir(context).createFile(mimeType,"testFile1").getUri();
    List<Uri> fileUris = Arrays.asList(testFile1Uri, testFile2Uri);
    Stream.of(fileUris).forEach(this::assertFileScheme);
    Intent intent = ShareUtils.buildIntent(fileUris, mimeType, null);
    List<Uri> sharedUris = intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
    Stream.of(sharedUris).forEach(this::assertContentScheme);
  }

  private void assertFileScheme(Uri uri) {
    assertScheme(uri, "file");
  }

  private void assertContentScheme(Uri uri) {
    assertScheme(uri, "content");
  }

  private void assertScheme(Uri uri, String scheme) {
    assertEquals(scheme, uri.getScheme());
  }
}

File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/util/ShareUtilsTest.kt
package org.totschnig.myexpenses.test.util

import android.content.Intent
import android.net.Uri
import androidx.core.content.IntentCompat
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.google.common.truth.Truth
import org.junit.Test
import org.junit.runner.RunWith
import org.totschnig.myexpenses.util.AppDirHelper
import org.totschnig.myexpenses.viewmodel.BaseFunctionalityViewModel

@RunWith(AndroidJUnit4::class)
class ShareUtilsTest {
    @Test
    fun shouldConvertSingleFileUri() {
        val mimeType = "text/plain"
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val appDir = AppDirHelper.getAppDir(context).getOrThrow()
        val testFile = appDir.createFile(mimeType, "testFile")!!
        val testFileUri = testFile.uri
        assertFileScheme(testFileUri)
        val fileUris = listOf(testFileUri)
        val intent = BaseFunctionalityViewModel.buildIntent(context, fileUris, mimeType, null)
        val sharedUris = IntentCompat.getParcelableArrayListExtra(intent, Intent.EXTRA_STREAM, Uri::class.java)!!
        assertContentScheme(sharedUris.first())
        //cleanup
        testFile.delete()
    }

    @Test
    fun shouldConvertMultipleFileUris() {
        val mimeType = "text/plain"
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val appDir = AppDirHelper.getAppDir(context).getOrThrow()
        val testFile1 = appDir.createFile(mimeType, "testFile1")!!
        val testFile2 = appDir.createFile(mimeType, "testFile2")!!
        val testFile1Uri = testFile1.uri
        val testFile2Uri = testFile2.uri
        val fileUris = listOf(testFile1Uri, testFile2Uri)
        fileUris.forEach { uri: Uri -> assertFileScheme(uri) }
        val intent = BaseFunctionalityViewModel.buildIntent(context, fileUris, mimeType, null)
        val sharedUris = IntentCompat.getParcelableArrayListExtra(intent, Intent.EXTRA_STREAM, Uri::class.java)!!
        sharedUris.forEach { uri: Uri -> assertContentScheme(uri) }
        //cleanup
        testFile1.delete()
        testFile2.delete()
    }

    private fun assertFileScheme(uri: Uri) {
        assertScheme(uri, "file")
    }

    private fun assertContentScheme(uri: Uri) {
        assertScheme(uri, "content")
    }

    private fun assertScheme(uri: Uri, scheme: String) {
        Truth.assertThat(uri.scheme).isEqualTo(scheme)
    }
}